channeldb/kvdb/etcd/readwrite_bucket.go

```diff
@@ -20,7 +21,12 @@ type readWriteBucket struct {
 
 // newReadWriteBucket creates a new rw bucket with the passed transaction
 // and bucket id.
-func newReadWriteBucket(tx *readWriteTx, id []byte) *readWriteBucket {
+func newReadWriteBucket(tx *readWriteTx, key, id []byte) *readWriteBucket {
+	if !bytes.Equal(id, rootBucketID()) {
+		// Add the bucket key/value to the lock set.
+		tx.lock(string(key), string(id))
```

Comment:
This is cool! It's basically like tree locking.

Reply:
yeah, this locks the keys along the bucket path, (although possibly creates more conflicts than simply locking the on the keys in the bucket but is also more correct)